### ISSUE TITLE
About Devito and Pytorch

### ISSUE THREAD

**Issue by StarrMoonn** [DIG-Kaust/Devito-fwi#25]
Hello！
Is this project still under development?

I'd like to use Devito to customize a wave equation and use some complex error functions, such as W1. However, Devito seems to require me to manually derive the gradient of W1 with respect to the model. Can I use PyTorch's automatic differentiation to connect this? It's similar to Deepwave, but Deepwave doesn't support customized wave equations.

**Comment by mrava87** [comment]
Hi @StarrMoonn,
yes, the project is still under development although not very active.

This example should you an idea on how to couple your own loss function into any propagator (here the acoustic one): https://github.com/DIG-Kaust/Devito-fwi/blob/main/notebooks/acoustic/AcousticVel_L2torch_1stage.ipynb

Here we just wrote a simple L2 norm with torch to show how you can use AD to get the adjoint source but of course you can implement any complicated loss you would like: https://github.com/DIG-Kaust/Devito-fwi/blob/28092ed2610d5432b09660d523521a0148df7898/devitofwi/loss/l2.py#L77

And feel free to contribute any of your loss functions if you want 😀

**Comment by StarrMoonn** [comment]
> 你好[@StarrMoonn](https://github.com/StarrMoonn)是的， 该项目虽然不是很活跃，但仍在开发中。
> 
> 这个例子应该能让你了解如何将自己的损失函数耦合到任何传播器（这里是声学传播器）：https://github.com/DIG-Kaust/Devito-fwi/blob/main/notebooks/acoustic/AcousticVel_L2torch_1stage.ipynb
> 
> 这里我们只是用 torch 编写了一个简单的 L2 范式来展示如何使用 AD 来获取伴随源，但当然你可以实现任何你想要的复杂损失：
> 
> [Devito-fwi/devitofwi/loss/l2.py](https://github.com/DIG-Kaust/Devito-fwi/blob/28092ed2610d5432b09660d523521a0148df7898/devitofwi/loss/l2.py#L77)
> 
> 第 77 行 [28092ed](/DIG-Kaust/Devito-fwi/commit/28092ed2610d5432b09660d523521a0148df7898)
> 
>  类 L2Torch（非线性算子）：
> 如果你愿意的话，欢迎贡献你的任何损失函数😀

Yet，I have seen the notebook https://github.com/DIG-Kaust/Devito-fwi/blob/28092ed2610d5432b09660d523521a0148df7898/devitofwi/loss/l2.py#L18，I discovered that this project can indeed inject custom loss into Devito based on the Pytorch implementation. Because my programming skills are relatively weak, I am currently focusing on FWI. I used to use deepwave, but it only encapsulates the same properties of acoustic and elastic waves and does not support custom wave equations. For example, if I want to study viscoacoustics, it is very difficult. I think Devito can help me a lot, but my goal is still to focus on the error function. AD can handle gradient backpropagation very well. I see that deepwave implements it through loss.backward, but this seems to be a module encapsulated by the original author. I don't understand it very well. I am looking for a combination of Devito and Pytorch. I have seen L2Torch and roughly understand the connection logic.
Therefore, as long as I customize the wave equation based on Devito and encapsulate the interface, similar to the class AcousticWave2D, https://github.com/DIG-Kaust/Devito-fwi/blob/28092ed2610d5432b09660d523521a0148df7898/devitofwi/waveengine/acoustic.py#L28, and then port the loss adapted to DeepWave to L2 Torch according to L2Torch, I can achieve a hybrid of Devito's custom wave equation and custom error function, right? For example, I want to implement viscoacoustic FWI + W1 Loss

**Comment by StarrMoonn** [comment]
Wow，[Matteo Ravasi](https://github.com/mrava87)，Your research covers a wide range, and your Marchenko is excellent. I still use your Volve Synthetic Data frequently in FWI. I've looked at your projects, and you're also developing Pylops. I saw a while ago that Pylops also supports FWI, but I've always been confused. With so many FWI tools available, does it have an advantage over Devito and JUDI? I see they seem to specialize in different areas of detail. I'm not a huge FWI expert, but I am a PhD student in geophysics, haha.

**Comment by mrava87** [comment]
Thanks @StarrMoonn 😀

No PyLops does not provide any interface to FWI as FWI is a nonlinear problem and PyLops is focused on linear ones. This is why we decided to create Devito-FWI as we prefer not to alter the focus of the PyLops core library. We do however have some operators based on Devito for LSRTM as this is a linear problem and so falls within the scope.

I’m not sure why there are many different tools for FWI. Devito-FWI is similar in spirit to JUDI as both use DeVito as propagator and build an inversion framework around. Deepwave bundles everything together, something which I’m not a great fan of, as I prefer to separate the propagators from the inverse problem setting, but this is a personal preference and depending on your specific project there may be scenarios you like one approach over the other
